### ISSUE TITLE
executeCheck() gets SIGSEGV when command in hook is invalid, and the agent reports nothing

### ISSUE THREAD

**Issue by trombik** [sensu/sensu-go#3779]
## Expected Behavior

even when configuration is invalid, the agent should not killed by SIGSERV, and should reports the problem.

## Current Behavior

when `command` in hook is invalid, the agent gets SIGSEGV. and the problem is not reported back to user. the status of the check remains `success`.

```
May 21 17:36:23  sensu_agent[2773]: panic: runtime error: invalid memory address or nil pointer dereference
May 21 17:36:23  sensu_agent[2773]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x68 pc=0xe5e0dc]
May 21 17:36:23  sensu_agent[2773]: 
May 21 17:36:23  sensu_agent[2773]: goroutine 87 [running]:
May 21 17:36:23  sensu_agent[2773]: github.com/sensu/sensu-go/agent.(*Agent).ExecuteHooks(0xc0001b6dc0, 0x128b0a0, 0xc00015c540, 0xc0005fc080, 0xc0000f8a80, 0x0, 0x6, 0x8, 0x0)
May 21 17:36:23  sensu_agent[2773]:     /wrkdirs/usr/ports/sysutils/sensu-go-agent/work/sensu-go-5.20.0/agent/hook.go:30 +0x26c
May 21 17:36:23  sensu_agent[2773]: github.com/sensu/sensu-go/agent.(*Agent).executeCheck(0xc0001b6dc0, 0x128b0a0, 0xc00015c540, 0xc0005fc080, 0xc000256000)
May 21 17:36:23  sensu_agent[2773]:     /wrkdirs/usr/ports/sysutils/sensu-go-agent/work/sensu-go-5.20.0/agent/check_handler.go:252 +0xe8f
May 21 17:36:23  sensu_agent[2773]: created by github.com/sensu/sensu-go/agent.(*Agent).handleCheck
May 21 17:36:23  sensu_agent[2773]:     /wrkdirs/usr/ports/sysutils/sensu-go-agent/work/sensu-go-5.20.0/agent/check_handler.go:69 +0x2c8
```

the end result is that, after the hook execution error, the check reports nothing. and the web ui says nothing about the problem.

## Possible Solution

whenever child process is killed, the agent should report the problem.

## Steps to Reproduce (for bugs)

1. create a hook with command including shell pipe, like this:

```yaml
type: HookConfig
api_version: core/v2
metadata:
  created_by: admin
  name: dmesg
  namespace: default
spec:
  command: dmesg | tail
  runtime_assets: null
  stdin: false
  timeout: 10
```

2. attach the hook to a check

## Context

fixing "nil pointer dereference" that caused SIGSEGV is one thing, but the agent should have not been fragile like this.

I would argue that using pipe is one of common operations for which the hook is designed. but that would be reserved for another bug report.

## Your Environment

* Sensu version used (sensuctl, sensu-backend, and/or sensu-agent): 5.20.0
* Installation method (packages, binaries, docker etc.): my own build
* Operating System and version (e.g. Ubuntu 14.04): FreeBSD 12.1

**Comment by trombik** [comment]
if you want to run a hook that includes pipe, the solution is:

```yaml
type: HookConfig
api_version: core/v2
metadata:
  created_by: admin
  name: dmesg
  namespace: default
spec:
  command: sh -c "dmesg | tail"
  runtime_assets: null
  stdin: false
  timeout: 10
```

**Comment by echlebek** [comment]
Thanks, the agent is definitely in need of some improvement here.

**Comment by echlebek** [comment]
As you say in the issue description, it's one thing to fix the segfault, but another to define a protocol for conveying failure to the backend. It's not clear that a failed hook should constitute a failed check if the hook is missing for instance. It seems that there is some design debt here that has to be accounted for.

**Comment by portertech** [comment]
Is the failed hook execution not captured in the check hook output and status? 

https://github.com/sensu/sensu-go/blob/16392dbff4ee2513ac1bcf7312a10eb862f5f51d/api/core/v2/hook.proto#L53-L56

As a check hook execution is a separate action, I don't think a hook failure should influence the check execution status.

**Comment by echlebek** [comment]
If a check fails to execute, an adhoc event is created and sent to the backend.

https://github.com/sensu/sensu-go/blob/master/agent/hook.go#L162

It's a little hard to follow because error handling is not well modeled.

The reason for the segfault is that the hook could not be resolved at all, and the function `getHookConfig` returns a nil pointer. It shouldn't have anything to do with the command being wrong, but I could be mistaken.

The question is: should we send a failure message when the hook cannot be resolved?

**Comment by trombik** [comment]
@echlebek is right, and i was wrong. the issue has nothing to do with `command`. it looks to me that `sensu-go` does not check dependencies of resources.

i replaced one of hooks with one that does not exist (`foobar`).  surprisingly (to me), the backend accept check definition with a hook that does not exist.

```yaml
type: CheckConfig
api_version: core/v2
metadata:
  created_by: admin
  name: check_http
  namespace: default
spec:
  check_hooks:
  - critical:
    - process_tree
    - foobar
    - syslog
  command: '{{ .labels.nagios_plugin_dir }}/check_http -H {{ index .labels "nginx_hostname"
    }}'
  env_vars: null
  handlers:
  - slack
  high_flap_threshold: 0
  interval: 60
  low_flap_threshold: 0
  output_metric_format: nagios_perfdata
  output_metric_handlers: null
  proxy_entity_name: ""
  publish: true
  round_robin: false
  runtime_assets: null
  secrets: null
  stdin: false
  subdue: null
  subscriptions:
  - web_server
  timeout: 0
  ttl: 0
```

and the same segfault.

```
May 22 07:15:27  sensu_agent[7947]: panic: runtime error: invalid memory address or nil pointer dereference
May 22 07:15:27  sensu_agent[7947]: [signal SIGSEGV: segmentation violation code=0x1 addr=0x68 pc=0xe5e0dc]
```

> Is the failed hook execution not captured in the check hook output and status?

to me as a user, no. the check status page shows stale status, `success`, when the monitored process is not running.

**Comment by echlebek** [comment]
It's true that in general sensu-go does not check resource dependencies. This is done in part to make it easier to dump and recreate configuration, and for the sake of storage layout simplicity.

I think that you should probably get some error notification when a hook does not exist, rather than silent failure. I'll modify the agent to send a failure even when the hook couldn't be looked up.

I have some confusion over the kind of failure even that's being sent right now. I could be wrong but it doesn't seem to have a check name. That's not valid to have, so unless it's getting set somewhere else, there could be an issue there as well.

More to come next week.

**Comment by trombik** [comment]
> It's true that in general sensu-go does not check resource dependencies. This is done in part to make it easier to dump and recreate configuration, and for the sake of storage layout simplicity.

is it a documented feature? which implies another feature request, possibly in another issue: as a user, I would like to make sure the running configuration has valid dependencies. as `sensu-go` does not have a configuration files for resources, there is no way to validate the configuration before applying a change. if you cannot validate the change when you add something, then there should be a way to validate the running configuration instead of waiting for agents to report errors.

> More to come next week.

that's good to know.

**Comment by echlebek** [comment]
This will be fixed in 5.21.

@trombik, regarding checking configuration dependencies, sensu-go is not likely to support this type of dependency checking any time soon, as the configuration is stored in a key-value store, not a relational database.
